File: tna/centralities.py
"""Centrality measures for TNA package.

Provides exact numerical equivalence to R TNA package centrality measures.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

import numpy as np
import pandas as pd
import networkx as nx
from scipy import linalg

if TYPE_CHECKING:
    from .model import TNA


# Available centrality measures
AVAILABLE_MEASURES = [
    'OutStrength',
    'InStrength',
    'ClosenessIn',
    'ClosenessOut',
    'Closeness',
    'Betweenness',
    'BetweennessRSP',
    'Diffusion',
    'Clustering'
]


def centralities(
    model: 'TNA',
    loops: bool = False,
    normalize: bool = False,
    measures: list[str] | None = None
) -> pd.DataFrame:
    """Compute centrality measures for a TNA model.

    Parameters
    ----------
    model : TNA
        A TNA model object
    loops : bool
        If True, include self-loops in calculations
    normalize : bool
        If True, normalize measures to [0, 1] using min-max normalization
    measures : list of str, optional
        Which measures to compute. If None, computes all available measures.
        Available: OutStrength, InStrength, ClosenessIn, ClosenessOut,
        Closeness, Betweenness, BetweennessRSP, Diffusion, Clustering

    Returns
    -------
    pd.DataFrame
        DataFrame with states as rows and centrality measures as columns
    """
    if measures is None:
        measures = AVAILABLE_MEASURES.copy()

    # Validate measures
    invalid = set(measures) - set(AVAILABLE_MEASURES)
    if invalid:
        raise ValueError(f"Unknown measures: {invalid}. Available: {AVAILABLE_MEASURES}")

    weights = model.weights.copy()
    n = weights.shape[0]

    # Remove self-loops if requested (R: diag(x) <- ifelse_(loops, diag(x), 0))
    if not loops:
        np.fill_diagonal(weights, 0)

    results = {}

    # Create igraph-compatible graph for NetworkX
    # R uses igraph, we use networkx but match behavior
    G = _create_graph(weights)

    # Compute requested measures in order they appear in AVAILABLE_MEASURES
    for measure in AVAILABLE_MEASURES:
        if measure not in measures:
            continue
        if measure == 'OutStrength':
            results['OutStrength'] = _out_strength(weights)
        elif measure == 'InStrength':
            results['InStrength'] = _in_strength(weights)
        elif measure == 'ClosenessIn':
            results['ClosenessIn'] = _closeness_in(G, n)
        elif measure == 'ClosenessOut':
            results['ClosenessOut'] = _closeness_out(G, n)
        elif measure == 'Closeness':
            results['Closeness'] = _closeness_all(G, n)
        elif measure == 'Betweenness':
            results['Betweenness'] = _betweenness(G, n)
        elif measure == 'BetweennessRSP':
            results['BetweennessRSP'] = _betweenness_rsp(weights)
        elif measure == 'Diffusion':
            results['Diffusion'] = _diffusion(weights)
        elif measure == 'Clustering':
            results['Clustering'] = _clustering(weights)

    # Create DataFrame
    df = pd.DataFrame(results, index=model.labels)

    # Normalize if requested (R: ranger function - min-max normalization)
    if normalize:
        for col in df.columns:
            col_min = df[col].min()
            col_max = df[col].max()
            if col_max > col_min:
                df[col] = (df[col] - col_min) / (col_max - col_min)
            else:
                df[col] = 0.0

    return df


def _create_graph(weights: np.ndarray) -> nx.DiGraph:
    """Create NetworkX DiGraph from weight matrix."""
    n = weights.shape[0]
    G = nx.DiGraph()
    G.add_nodes_from(range(n))
    for i in range(n):
        for j in range(n):
            if weights[i, j] > 0:
                G.add_edge(i, j, weight=weights[i, j])
    return G


def _out_strength(weights: np.ndarray) -> np.ndarray:
    """Compute out-strength (sum of outgoing edge weights).

    R equivalent: igraph::strength(g, mode = "out")
    """
    return weights.sum(axis=1)


def _in_strength(weights: np.ndarray) -> np.ndarray:
    """Compute in-strength (sum of incoming edge weights).

    R equivalent: igraph::strength(g, mode = "in")
    """
    return weights.sum(axis=0)


def _closeness_in(G: nx.DiGraph, n: int) -> np.ndarray:
    """Compute incoming closeness centrality.

    R equivalent: igraph::closeness(g, mode = "in")
    igraph uses: (n-1) / sum(distances_to_node)
    """
    result = np.zeros(n)
    # Reverse graph to compute paths TO each node
    rev_G = G.reverse()

    for i in range(n):
        try:
            lengths = nx.single_source_dijkstra_path_length(
                rev_G, i, weight=lambda u, v, d: 1.0 / d['weight']
            )
            # Sum of distances from all reachable nodes
            total_dist = sum(d for j, d in lengths.items() if j != i)
            n_reachable = len([j for j in lengths if j != i])
            if total_dist > 0:
                result[i] = n_reachable / total_dist
        except (nx.NetworkXError, nx.NodeNotFound):
            result[i] = 0.0

    return result


def _closeness_out(G: nx.DiGraph, n: int) -> np.ndarray:
    """Compute outgoing closeness centrality.

    R equivalent: igraph::closeness(g, mode = "out")
    """
    result = np.zeros(n)

    for i in range(n):
        try:
            lengths = nx.single_source_dijkstra_path_length(
                G, i, weight=lambda u, v, d: 1.0 / d['weight']
            )
            total_dist = sum(d for j, d in lengths.items() if j != i)
            n_reachable = len([j for j in lengths if j != i])
            if total_dist > 0:
                result[i] = n_reachable / total_dist
        except (nx.NetworkXError, nx.NodeNotFound):
            result[i] = 0.0

    return result


def _closeness_all(G: nx.DiGraph, n: int) -> np.ndarray:
    """Compute overall closeness centrality (mode = "all").

    R equivalent: igraph::closeness(g, mode = "all")
    Treats graph as undirected for distance calculation.
    """
    result = np.zeros(n)

    # Create undirected version
    U = G.to_undirected()

    for i in range(n):
        try:
            lengths = nx.single_source_dijkstra_path_length(
                U, i, weight=lambda u, v, d: 1.0 / d['weight']
            )
            total_dist = sum(d for j, d in lengths.items() if j != i)
            n_reachable = len([j for j in lengths if j != i])
            if total_dist > 0:
                result[i] = n_reachable / total_dist
        except (nx.NetworkXError, nx.NodeNotFound):
            result[i] = 0.0

    return result


def _betweenness(G: nx.DiGraph, n: int) -> np.ndarray:
    """Compute betweenness centrality.

    R equivalent: igraph::betweenness(g)
    Uses weighted shortest paths.
    """
    # igraph uses 1/weight as distance by default for weighted graphs
    bc = nx.betweenness_centrality(
        G,
        weight=lambda u, v, d: 1.0 / d['weight'] if d['weight'] > 0 else float('inf'),
        normalized=False
    )
    return np.array([bc.get(i, 0.0) for i in range(n)])


def _betweenness_rsp(weights: np.ndarray, beta: float = 0.01) -> np.ndarray:
    """Compute Randomized Shortest Path betweenness centrality.

    Exact port of R tna::rsp_bet function (Kivimäki et al. 2016).

    R code:
        n <- ncol(mat)
        D <- .rowSums(mat, m = n, n = n)
        if (any(D == 0)) return(NA)
        P_ref <- diag(D^-1) %*% mat
        C <- mat^-1
        C[is.infinite(C)] <- 0
        W <- P_ref * exp(-beta * C)
        Z <- solve(diag(1, n, n) - W)
        Z_recip <- Z^-1
        Z_recip[is.infinite(Z_recip)] <- 0
        Z_recip_diag <- diag(Z_recip) * diag(1, n, n)
        out <- diag(tcrossprod(Z, Z_recip - n * Z_recip_diag) %*% Z)
        out <- round(out)
        out <- out - min(out) + 1
    """
    n = weights.shape[0]
    mat = weights.copy()

    # D <- .rowSums(mat, m = n, n = n)
    D = mat.sum(axis=1)

    # if (any(D == 0)) return(NA)
    if np.any(D == 0):
        return np.full(n, np.nan)

    # P_ref <- diag(D^-1) %*% mat
    P_ref = np.diag(1.0 / D) @ mat

    # C <- mat^-1; C[is.infinite(C)] <- 0
    with np.errstate(divide='ignore'):
        C = 1.0 / mat
    C[np.isinf(C)] = 0

    # W <- P_ref * exp(-beta * C)
    W = P_ref * np.exp(-beta * C)

    # Z <- solve(diag(1, n, n) - W)
    try:
        Z = linalg.inv(np.eye(n) - W)
    except linalg.LinAlgError:
        return np.full(n, np.nan)

    # Z_recip <- Z^-1; Z_recip[is.infinite(Z_recip)] <- 0
    with np.errstate(divide='ignore'):
        Z_recip = 1.0 / Z
    Z_recip[np.isinf(Z_recip)] = 0

    # Z_recip_diag <- diag(Z_recip) * diag(1, n, n)
    Z_recip_diag = np.diag(np.diag(Z_recip))

    # out <- diag(tcrossprod(Z, Z_recip - n * Z_recip_diag) %*% Z)
    # tcrossprod(A, B) = A %*% t(B)
    out = np.diag(Z @ (Z_recip - n * Z_recip_diag).T @ Z)

    # out <- round(out)
    out = np.round(out)

    # out <- out - min(out) + 1
    out = out - out.min() + 1

    return out


def _diffusion(weights: np.ndarray) -> np.ndarray:
    """Compute diffusion centrality.

    Exact port of R tna::diffusion function (Banerjee et al. 2014).

    R code:
        s <- 0
        n <- ncol(mat)
        p <- diag(1, n, n)
        for (i in seq_len(n)) {
            p <- p %*% mat
            s <- s + p
        }
        .rowSums(s, n, n)
    """
    n = weights.shape[0]
    mat = weights.copy()

    s = np.zeros((n, n))
    p = np.eye(n)

    for _ in range(n):
        p = p @ mat
        s = s + p

    # .rowSums(s, n, n) - row sums
    return s.sum(axis=1)


def _clustering(weights: np.ndarray) -> np.ndarray:
    """Compute signed clustering coefficient.

    Exact port of R tna::wcc function (Zhang and Horvath 2005).
    Called on symmetric matrix: wcc(x + t(x))

    R code:
        wcc <- function(mat) {
            diag(mat) <- 0
            n <- ncol(mat)
            num <- diag(mat %*% mat %*% mat)
            den <- .colSums(mat, n, n)^2 - .colSums(mat^2, n, n)
            num / den
        }
        # Called as: wcc(x + t(x))
    """
    # Symmetrize: x + t(x)
    mat = weights + weights.T

    # diag(mat) <- 0
    np.fill_diagonal(mat, 0)

    n = mat.shape[0]

    # num <- diag(mat %*% mat %*% mat)
    num = np.diag(mat @ mat @ mat)

    # den <- .colSums(mat, n, n)^2 - .colSums(mat^2, n, n)
    col_sums = mat.sum(axis=0)
    col_sums_sq = (mat ** 2).sum(axis=0)
    den = col_sums ** 2 - col_sums_sq

    # num / den (handle division by zero)
    with np.errstate(divide='ignore', invalid='ignore'):
        result = num / den
    result[~np.isfinite(result)] = 0.0

    return result

File: tna-js/src/analysis/centralities.ts
/**
 * Centrality measures for TNA.
 * Port of Python tna/centralities.py
 *
 * Uses hand-rolled graph algorithms instead of graphology to match
 * the exact R TNA behavior (igraph with 1/weight distances).
 */
import { Matrix } from '../core/matrix.js';
import type { TNA, GroupTNA, CentralityMeasure, CentralityResult } from '../core/types.js';
import { isGroupTNA, groupEntries } from '../core/group.js';

export const AVAILABLE_MEASURES: CentralityMeasure[] = [
  'OutStrength', 'InStrength', 'ClosenessIn', 'ClosenessOut',
  'Closeness', 'Betweenness', 'BetweennessRSP', 'Diffusion', 'Clustering',
];

/**
 * Compute centrality measures for a TNA model.
 */
export function centralities(
  model: TNA | GroupTNA,
  options?: {
    loops?: boolean;
    normalize?: boolean;
    measures?: CentralityMeasure[];
  },
): CentralityResult {
  // Handle GroupTNA
  if (isGroupTNA(model)) {
    const allLabels: string[] = [];
    const allGroups: string[] = [];
    const allMeasures: Record<CentralityMeasure, number[]> = {} as Record<CentralityMeasure, number[]>;

    for (const [name, m] of groupEntries(model)) {
      const result = centralities(m, options);
      for (let i = 0; i < result.labels.length; i++) {
        allLabels.push(result.labels[i]!);
        allGroups.push(name);
      }
      for (const [measure, values] of Object.entries(result.measures) as [CentralityMeasure, Float64Array][]) {
        if (!allMeasures[measure]) allMeasures[measure] = [];
        for (let i = 0; i < values.length; i++) {
          allMeasures[measure]!.push(values[i]!);
        }
      }
    }

    const measures: Record<CentralityMeasure, Float64Array> = {} as Record<CentralityMeasure, Float64Array>;
    for (const [m, vals] of Object.entries(allMeasures) as [CentralityMeasure, number[]][]) {
      measures[m] = new Float64Array(vals);
    }

    return { labels: allLabels, measures, groups: allGroups };
  }

  const tnaModel = model as TNA;
  const requestedMeasures = options?.measures ?? [...AVAILABLE_MEASURES];
  const loops = options?.loops ?? false;
  const normalize = options?.normalize ?? false;

  const weights = tnaModel.weights.clone();
  const n = weights.rows;

  // Remove self-loops if not requested
  if (!loops) {
    for (let i = 0; i < n; i++) weights.set(i, i, 0);
  }

  const measures: Record<string, Float64Array> = {};

  for (const measure of AVAILABLE_MEASURES) {
    if (!requestedMeasures.includes(measure)) continue;

    switch (measure) {
      case 'OutStrength':
        measures.OutStrength = outStrength(weights);
        break;
      case 'InStrength':
        measures.InStrength = inStrength(weights);
        break;
      case 'ClosenessIn':
        measures.ClosenessIn = closenessIn(weights, n);
        break;
      case 'ClosenessOut':
        measures.ClosenessOut = closenessOut(weights, n);
        break;
      case 'Closeness':
        measures.Closeness = closenessAll(weights, n);
        break;
      case 'Betweenness':
        measures.Betweenness = betweenness(weights, n);
        break;
      case 'BetweennessRSP':
        measures.BetweennessRSP = betweennessRSP(weights);
        break;
      case 'Diffusion':
        measures.Diffusion = diffusion(weights);
        break;
      case 'Clustering':
        measures.Clustering = clustering(weights);
        break;
    }
  }

  // Normalize if requested (min-max per measure)
  if (normalize) {
    for (const key of Object.keys(measures)) {
      const vals = measures[key]!;
      let min = Infinity;
      let max = -Infinity;
      for (let i = 0; i < vals.length; i++) {
        if (vals[i]! < min) min = vals[i]!;
        if (vals[i]! > max) max = vals[i]!;
      }
      if (max > min) {
        for (let i = 0; i < vals.length; i++) {
          vals[i] = (vals[i]! - min) / (max - min);
        }
      } else {
        vals.fill(0);
      }
    }
  }

  return {
    labels: tnaModel.labels,
    measures: measures as Record<CentralityMeasure, Float64Array>,
  };
}

// ---- Strength ----

function outStrength(weights: Matrix): Float64Array {
  return weights.rowSums();
}

function inStrength(weights: Matrix): Float64Array {
  return weights.colSums();
}

// ---- Dijkstra shortest paths with 1/weight distances ----

/** Dijkstra from source, using 1/weight as distance. Returns distances to all nodes. */
function dijkstra(n: number, getWeight: (from: number, to: number) => number, source: number): Float64Array {
  const dist = new Float64Array(n).fill(Infinity);
  const visited = new Uint8Array(n);
  dist[source] = 0;

  for (let step = 0; step < n; step++) {
    // Find unvisited node with min distance
    let u = -1;
    let minDist = Infinity;
    for (let i = 0; i < n; i++) {
      if (!visited[i] && dist[i]! < minDist) {
        minDist = dist[i]!;
        u = i;
      }
    }
    if (u === -1) break;
    visited[u] = 1;

    for (let v = 0; v < n; v++) {
      if (visited[v]) continue;
      const w = getWeight(u, v);
      if (w > 0) {
        const d = 1 / w; // inverse weight as distance
        const newDist = dist[u]! + d;
        if (newDist < dist[v]!) {
          dist[v] = newDist;
        }
      }
    }
  }

  return dist;
}

// ---- Closeness ----

function closenessIn(weights: Matrix, n: number): Float64Array {
  const result = new Float64Array(n);

  for (let i = 0; i < n; i++) {
    // Reverse graph: paths TO node i
    const dist = dijkstra(n, (from, to) => weights.get(to, from), i);

    let totalDist = 0;
    for (let j = 0; j < n; j++) {
      if (j !== i && isFinite(dist[j]!)) totalDist += dist[j]!;
    }
    result[i] = totalDist > 0 ? 1 / totalDist : 0;
  }

  return result;
}

function closenessOut(weights: Matrix, n: number): Float64Array {
  const result = new Float64Array(n);

  for (let i = 0; i < n; i++) {
    const dist = dijkstra(n, (from, to) => weights.get(from, to), i);

    let totalDist = 0;
    for (let j = 0; j < n; j++) {
      if (j !== i && isFinite(dist[j]!)) totalDist += dist[j]!;
    }
    result[i] = totalDist > 0 ? 1 / totalDist : 0;
  }

  return result;
}

function closenessAll(weights: Matrix, n: number): Float64Array {
  // Mode="all" — undirected with max weight per edge pair
  const symWeights = Matrix.zeros(n, n);
  for (let i = 0; i < n; i++) {
    for (let j = i + 1; j < n; j++) {
      const w = Math.max(weights.get(i, j), weights.get(j, i));
      symWeights.set(i, j, w);
      symWeights.set(j, i, w);
    }
  }

  const result = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    const dist = dijkstra(n, (from, to) => symWeights.get(from, to), i);

    let totalDist = 0;
    for (let j = 0; j < n; j++) {
      if (j !== i && isFinite(dist[j]!)) totalDist += dist[j]!;
    }
    result[i] = totalDist > 0 ? 1 / totalDist : 0;
  }

  return result;
}

// ---- Betweenness (Brandes' algorithm with 1/weight distances) ----

function betweenness(weights: Matrix, n: number): Float64Array {
  const CB = new Float64Array(n);

  for (let s = 0; s < n; s++) {
    // Modified Dijkstra
    const stack: number[] = [];
    const pred: number[][] = Array.from({ length: n }, () => []);
    const sigma = new Float64Array(n);
    const dist = new Float64Array(n).fill(Infinity);

    sigma[s] = 1;
    dist[s] = 0;

    // Priority queue (simple linear scan for small n)
    const visited = new Uint8Array(n);

    for (let step = 0; step < n; step++) {
      let u = -1;
      let minDist = Infinity;
      for (let i = 0; i < n; i++) {
        if (!visited[i] && dist[i]! < minDist) {
          minDist = dist[i]!;
          u = i;
        }
      }
      if (u === -1) break;
      visited[u] = 1;
      stack.push(u);

      for (let v = 0; v < n; v++) {
        if (visited[v]) continue;
        const w = weights.get(u, v);
        if (w <= 0) continue;
        const d = 1 / w;
        const newDist = dist[u]! + d;

        if (newDist < dist[v]! - 1e-15) {
          dist[v] = newDist;
          sigma[v] = sigma[u]!;
          pred[v] = [u];
        } else if (Math.abs(newDist - dist[v]!) < 1e-15) {
          sigma[v] = sigma[v]! + sigma[u]!;
          pred[v]!.push(u);
        }
      }
    }

    // Back-propagation
    const delta = new Float64Array(n);
    while (stack.length > 0) {
      const w = stack.pop()!;
      for (const v of pred[w]!) {
        const frac = (sigma[v]! / sigma[w]!) * (1 + delta[w]!);
        delta[v] = delta[v]! + frac;
      }
      if (w !== s) {
        CB[w] = CB[w]! + delta[w]!;
      }
    }
  }

  return CB;
}

// ---- Randomized Shortest Path Betweenness (Kivimaki et al. 2016) ----

function betweennessRSP(weights: Matrix, beta = 0.01): Float64Array {
  const n = weights.rows;
  const mat = weights.clone();

  // D <- rowSums(mat)
  const D = mat.rowSums();

  // if (any(D == 0)) return NA
  for (let i = 0; i < n; i++) {
    if (D[i] === 0) return new Float64Array(n).fill(NaN);
  }

  // P_ref <- diag(D^-1) %*% mat
  const Pref = Matrix.zeros(n, n);
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      Pref.set(i, j, mat.get(i, j) / D[i]!);
    }
  }

  // C <- mat^-1; C[is.infinite(C)] <- 0
  const C = mat.map((v) => (v === 0 ? 0 : 1 / v));

  // W <- P_ref * exp(-beta * C)
  const W = Matrix.zeros(n, n);
  for (let i = 0; i < n; i++) {
    for (let j = 0; j < n; j++) {
      W.set(i, j, Pref.get(i, j) * Math.exp(-beta * C.get(i, j)));
    }
  }

  // Z <- solve(I - W)
  const IminusW = Matrix.eye(n).sub(W);
  let Z: Matrix;
  try {
    Z = IminusW.inverse();
  } catch {
    return new Float64Array(n).fill(NaN);
  }

  // Z_recip <- Z^-1; Z_recip[is.infinite(Z_recip)] <- 0
  const Zrecip = Z.map((v) => {
    if (v === 0) return 0;
    const inv = 1 / v;
    return isFinite(inv) ? inv : 0;
  });

  // Z_recip_diag <- diag(Z_recip) * I
  const ZrecipDiag = Matrix.diag(Zrecip.diag());

  // out <- diag(Z %*% (Z_recip - n * Z_recip_diag)^T %*% Z)
  // = diag(Z @ (Zrecip - n*ZrecipDiag).T @ Z)
  const inner = Zrecip.sub(ZrecipDiag.scale(n));
  const step1 = Z.matmul(inner.transpose());
  const step2 = step1.matmul(Z);

  const out = step2.diag();

  // Round
  for (let i = 0; i < n; i++) {
    out[i] = Math.round(out[i]!);
  }

  // out = out - min(out) + 1
  let minVal = Infinity;
  for (let i = 0; i < n; i++) {
    if (out[i]! < minVal) minVal = out[i]!;
  }
  for (let i = 0; i < n; i++) {
    out[i] = out[i]! - minVal + 1;
  }

  return out;
}

// ---- Diffusion centrality (Banerjee et al. 2014) ----

function diffusion(weights: Matrix): Float64Array {
  const n = weights.rows;
  let s = Matrix.zeros(n, n);
  let p = Matrix.eye(n);

  for (let i = 0; i < n; i++) {
    p = p.matmul(weights);
    s = s.add(p);
  }

  return s.rowSums();
}

// ---- Clustering coefficient (Zhang and Horvath 2005) ----

function clustering(weights: Matrix): Float64Array {
  // Symmetrize: mat = weights + weights.T
  const mat = weights.add(weights.transpose());

  // diag(mat) <- 0
  for (let i = 0; i < mat.rows; i++) mat.set(i, i, 0);

  const n = mat.rows;

  // num = diag(mat %*% mat %*% mat)
  const mat2 = mat.matmul(mat);
  const mat3 = mat2.matmul(mat);
  const num = mat3.diag();

  // den = colSums(mat)^2 - colSums(mat^2)
  const colSums = mat.colSums();
  const matSq = mat.map((v) => v * v);
  const colSumsSq = matSq.colSums();

  const result = new Float64Array(n);
  for (let i = 0; i < n; i++) {
    const den = colSums[i]! * colSums[i]! - colSumsSq[i]!;
    result[i] = den !== 0 ? num[i]! / den : 0;
  }

  return result;
}

/**
 * Compute edge betweenness centrality and return as a new TNA model.
 */
export function betweennessNetwork(model: TNA | GroupTNA): TNA | Record<string, TNA> {
  if (isGroupTNA(model)) {
    const result: Record<string, TNA> = {};
    for (const [name, m] of groupEntries(model)) {
      result[name] = betweennessNetwork(m) as TNA;
    }
    return result;
  }

  const tnaModel = model as TNA;
  const weights = tnaModel.weights;
  const n = weights.rows;

  // Use Brandes' algorithm but collect edge betweenness
  const edgeBet = Matrix.zeros(n, n);

  for (let s = 0; s < n; s++) {
    const stack: number[] = [];
    const pred: number[][] = Array.from({ length: n }, () => []);
    const sigma = new Float64Array(n);
    const dist = new Float64Array(n).fill(Infinity);

    sigma[s] = 1;
    dist[s] = 0;

    const visited = new Uint8Array(n);

    for (let step = 0; step < n; step++) {
      let u = -1;
      let minDist = Infinity;
      for (let i = 0; i < n; i++) {
        if (!visited[i] && dist[i]! < minDist) {
          minDist = dist[i]!;
          u = i;
        }
      }
      if (u === -1) break;
      visited[u] = 1;
      stack.push(u);

      for (let v = 0; v < n; v++) {
        if (visited[v]) continue;
        const w = weights.get(u, v);
        if (w <= 0) continue;
        const d = 1 / w;
        const newDist = dist[u]! + d;

        if (newDist < dist[v]! - 1e-15) {
          dist[v] = newDist;
          sigma[v] = sigma[u]!;
          pred[v] = [u];
        } else if (Math.abs(newDist - dist[v]!) < 1e-15) {
          sigma[v] = sigma[v]! + sigma[u]!;
          pred[v]!.push(u);
        }
      }
    }

    const delta = new Float64Array(n);
    while (stack.length > 0) {
      const w = stack.pop()!;
      for (const v of pred[w]!) {
        const frac = (sigma[v]! / sigma[w]!) * (1 + delta[w]!);
        delta[v] = delta[v]! + frac;
        edgeBet.set(v, w, edgeBet.get(v, w) + sigma[v]! / sigma[w]! * (1 + delta[w]!));
      }
    }
  }

  return {
    weights: edgeBet,
    inits: new Float64Array(tnaModel.inits),
    labels: [...tnaModel.labels],
    data: tnaModel.data,
    type: 'betweenness',
    scaling: [...tnaModel.scaling],
  };
}
